src/snprintf.c

```diff
@@ -39,6 +39,8 @@ rcutils_snprintf(char * buffer, size_t buffer_size, const char * format, ...)
 int
 rcutils_vsnprintf(char * buffer, size_t buffer_size, const char * format, va_list args)
 {
+  RCUTILS_CAN_RETURN_WITH_ERROR_OF(-1);
```

Comment:
The main concern here is that we are going to litter our code with these macros.  It's not the end of the world, but it is noisy and can make the code harder to read.

I have 2 separate suggestions:
1.  Before we start down this road, let's see what @hidmic's work on Mimick looks like.  If we can make that work, that will obviate the need for this, I think (I'm currently ignoring the fact that Mimick doesn't work for C++, while this one would).
1.  If we can't get Mimick to work for some reason, another alternative is to use the `LD_PRELOAD` mechanism to override the symbols as needed.  This is less desirable than Mimick since it is limited to Linux-only, but I think we can make the case (at least for now) that doing it on one platform is better than doing it on zero platforms.

The benefit to both solutions above is that there is no changes to the code, just changes to the tests.  But I'd also like to hear what others think.

Reply:
I pushed this PR last night, and intended to give a more full writeup. Please take a look at the above PR description for some more information as well as possible extensions.

The approach taken in this PR allows for immediate cross-platform, cross-language utilization with simple unit tests with no external dependencies. You can write a unit test that easily verifies your code is hardened against failures in upstream libraries. See https://github.com/ros2/rcl_logging/pull/48 for an example.

Another approach was to add an `RCUTILS_CHECK_` macro that replaced simple if statements. Based on feedback from you and @wjwwood that also desired to see that we test the same code that we ship, I took this approach. The use of a macro like `RCUTILS_CAN_RETURN_WITH_ERROR_OF` allows for testing shipped production libraries against modified upstream libraries, which is not something our CI or buildfarm are setup for today but if desired could be feasibly updated in the future.

I took care in the design of `RCUTILS_CAN_RETURN_WITH_ERROR_OF` to ensure that it wasn't just litter. It actually serves an additional purpose of describing intended failure cases of the function. While docblocks typically try to capture the possible failures cases, this would be another approach that can also be integrated with unit testing. It may be difficult for a maintainer to see all the possible return values of a high level function that it passes on from lower level functions. However, with unit tests utilizing `RCUTILS_MAYBE_RETURN_ERROR`, you can verify that all possible return values are sufficiently understood.

This PR is intended as a starting point, and I want to make it as useful as possible. There are several possible feature additions that I think could fill it out, but I would like to get some more buy in as well.

Reply:
Thank you, it does help clarify your thinking. My hope is that this sort of feature is used in conjunction with mocking like what @hidmic is investigating. That sort of testing would allow for more targeted failures, while this approach especially with the for loop example might be more like throwing a grenade at your code and seeing if it survives.

You're absolutely right that there are a large number of places where the `RCUTILS_CAN_RETURN_WITH_ERROR_OF` macro can be placed, and it would be overwhelming to put them in every single location at the outset, or even for every contribution. But I also think it's possible to add this macro in piecemeal, and we can start by only adding it in when it's needed for unit tests. For example, to push rcl_logging_spdlog to 100% coverage, you only need `rcutils_snprintf` to fail, but given the correct inputs that function probably never will on our supported platforms.

My suggestion with this sort of change is to start small and expand it as its role in this codebase is better understood. That's partially the reason why I only introduce one flavor of `RCUTILS_CAN_RETURN_WITH_ERROR_OF`, even though I also want a variadic version.

That macro is empty if fault injection is not enabled, which I think helps make it easier to remove in the future. The example unit test in rcl_logging_spdlog would pass if `RCUTILS_CAN_RETURN_WITH_ERROR_OF` did not exist at all. We could also properly deprecate `RCUTILS_SET_FAULT_INJECTION_COUNT` to help point external developers to the appropriate replacement.

I also agree that seeing this macro at the top of a function would be strange and unusual to many developers, particularly newer ROS 2 users who are unfamiliar with the code. While I tried to be careful about the macro name, I'm also very open to rewording `RCUTILS_CAN_RETURN_WITH_ERROR_OF`. I chose the wording so that it's hopefully clear to developers unfamiliar with it that doesn't change the nominal behavior of the function, while conveying more information about the intent of the function's behavior if you are familiar with it.

Reply:
> Before we start down this road, let's see what @hidmic's work on Mimick looks like. If we can make that work, that will obviate the need for this, I think (I'm currently ignoring the fact that Mimick doesn't work for C++, while this one would).

@clalancette I think this serves a different use case. Mocking is good for white-box unit testing, while this is performing black-box integration testing to see how the system, with all its side effects, behaves in an exceptional event (akin to exception safety verification). Could we use mocks to implement it? Yes, but we'd end up coupling a package to its dependencies' implementation and recreating side effects -- possible, but much harder.

IMHO, `RCUTILS_CAN_RETURN_WITH_ERROR_OF`, though atypical, isn't strictly worse than the prologue of `RCUTILS_CHECK_ARGUMENT_FOR_NULL` our C functions usually have.